### PR TITLE
Batching requests

### PR DESCRIPTION
**Pull request by tiina303**
Didn't add tests yet & the same problem that Example likely shouldn't be where it is, but I didn't know how to get deps to work properly.

### CODE REVIEW

File: posthog/src/main/java/com/posthog/java/QueueManager.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        public synchronized void add(JSONObject json) {
+            ptr.add(json);
+        }
+
+        public synchronized List<JSONObject> retrieveAndReset() {
+            if (isEmpty()) {
+                return Collections.emptyList();
+            }
+            List<JSONObject> cur = ptr;
+            ptr = new LinkedList<JSONObject>();
```
Comment (yakkomajuri) [inline posthog/src/main/java/com/posthog/java/QueueManager.java:37]: Why are we mixing `LinkedList` with `List`? 

(side note: I haven't written Java in a while)

File: posthog/src/main/java/com/posthog/java/QueueManager.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    private void updateSendAfter() {
+        sendAfter = Instant.now().plusMillis(maxTimeToSendMs);
+    }
+
+    private void sleep() {
+        try {
+            Thread.sleep(sleepMs);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+            stop = true; // TODO: should we stop? should trigger an interrupt up?
```
Comment (yakkomajuri) [inline posthog/src/main/java/com/posthog/java/QueueManager.java:78]: Should be fine to just keep going? Although the stack trace might confuse people

File: posthog/src/main/java/com/posthog/java/PostHog.java [hunk 3]
```diff
@@ -43,30 +39,32 @@ public PostHog build() {
...
+            // TODO Auto-generated catch block
+            e.printStackTrace();
         }
     }
 
+    private void startQueueManager() {
+        queueManager = new QueueManager(sender);
+        queueManagerThread = new Thread(queueManager, "PostHog QueueManager thread");
+        queueManagerThread.start();
+        // TODO handle interrupts? (via addShutdownHook)
```
Comment (yakkomajuri) [inline posthog/src/main/java/com/posthog/java/PostHog.java:60]: yeah let's try to send on shutdown

**Review by yakkomajuri — COMMENTED** [review]
omw to leave the wework so partial comments

File: posthog/src/main/java/com/posthog/java/QueueManager.java (hunk 1 above)
Comment (tiina303) [reply posthog/src/main/java/com/posthog/java/QueueManager.java:37]: `List` is an interface (https://docs.oracle.com/javase/8/docs/api/java/util/List.html) so we can't create an object of `List` directly, but we have to pick one of the implementing classes, e.g. `LinkedList` to create an instance, however I do want this class to externally visibly operate on the `List` interface, so I can change the implementation without impacting any users, e.g. say I later want to change the `LinkedList` to  `ArrayList` I only need to make changes within this class.

File: posthog/src/main/java/com/posthog/java/PostHog.java (hunk 3 above)
Comment (tiina303) [reply posthog/src/main/java/com/posthog/java/PostHog.java:60]: I wasn't sure yet of all the implications of adding `addShutdownHook` to call `shutdown` & whether we want that, was going to look into it.

Comment (yakkomajuri) [reply posthog/src/main/java/com/posthog/java/PostHog.java:60]: what I'd like to see is just us trying to send requests if we come crashing down

**Review by yakkomajuri — APPROVED** [review]
approving so we can ship and iterate. 

let's ship this with a big alpha tag and disclaimer and see how it works.

then I'd love to see some docs and tests - probably docs first so we can easily try it and see

**Pull request merged**